test/foundry/deploy/03_Erc20Enforcer.t.sol

```diff
@@ -0,0 +1,697 @@
+  }
+
+  /*
+   * ERC20 TEST
+   *    Tokens Approvals
+   */
+
+  function testApporvalRevertZeroAddress() public {
+    vm.expectRevert("ERC20: spender is zero address");
+    holographERC20.approve(zeroAddress, maxValue);
+  }
+
+  function testApporvalEmitEvent() public {
+    vm.expectEmit(true, true, false, true);
+    emit Approval(deployer, alice, maxValue);
+    approvalToAlice(maxValue);
+  }
+
+  function testDecreaseAllowanceEmitEvent() public {
+    increaseAllowanceToAlice(maxValue);
+    vm.expectEmit(true, true, false, true);
+    emit Approval(deployer, alice, halfInverseValue);
+    decreaseAllowanceToAlice(halfValue);
+  }
+
+  function testDecreaseAllowanceBelongToZeroRevert() public {
+    increaseAllowanceToAlice(maxValue);
+    decreaseAllowanceToAlice(halfValue);
+    vm.expectRevert("ERC20: decreased below zero");
+    decreaseAllowanceToAlice(maxValue);
+  }
+
+  function testIncreaseAllowanceAboveToMaxValueRevert() public {
+    increaseAllowanceToAlice(maxValue);
+    vm.expectRevert("ERC20: increased above max value");
+    increaseAllowanceToAlice(maxValue);
+  }
+
+  function testDecreaseAllowanceToZero() public {
+    increaseAllowanceToAlice(maxValue);
+    vm.expectEmit(true, true, false, true);
+    emit Approval(deployer, alice, 0);
+    decreaseAllowanceToAlice(maxValue);
+  }
+
+  //same testApporvalEmitEvent
+  function testIncreaseAllowanceToMaxValue() public {
+    vm.expectEmit(true, true, false, true);
+    emit Approval(deployer, alice, maxValue);
+    increaseAllowanceToAlice(maxValue);
+  }
+
+  /*
+   * ERC20 TEST
+   *    Failed Transfers
+   */
+
+  function testTransferNotEnoughTokensRevert() public {
+    vm.expectRevert("ERC20: amount exceeds balance");
+    holographERC20.transfer(alice, maxValue);
+  }
+
+  function testTransferToZeroAddressRevert() public {
+    vm.expectRevert("ERC20: recipient is zero address");
+    holographERC20.transfer(zeroAddress, maxValue);
+  }
+
+  function testTransferFromZeroAddressRevert() public {
+    vm.expectRevert("ERC20: amount exceeds allowance");
+    holographERC20.transferFrom(zeroAddress, alice, maxValue);
+  }
+
+  function testTransferFromNotAprrovalAddressRevert() public {
+    vm.expectRevert("ERC20: amount exceeds allowance");
+    holographERC20.transferFrom(deployer, alice, maxValue);
+  }
+
+  function testTransferFromSmallerAprrovalAmountRevert() public {
+    approvalToBob(halfValue);
+    vm.expectRevert("ERC20: amount exceeds allowance");
+    vm.prank(bob);
+    holographERC20.transferFrom(deployer, alice, maxValue);
+  }
+
+  function testErc20RecivedNonContractRevert() public {
+    vm.expectRevert("ERC20: operator not contract");
+    holographERC20.onERC20Received(
+      deployer,
+      deployer,
+      initialValue,
+      bytes(abi.encode("0x0000000000000000000000000000000000000000"))
+    );
+  }
+
+  //TODO see why mock token have balance? remove Fail to the name of the function
```

Comment:
question: is this TODO still pending? if not, let's remove it

Reply:
yes, if it is part of the skip, which is also in hardhat.
When looking for the bug it was found to be balanced, hence the comment.